### ISSUE TITLE
processing for metadata "halts" the extension

### ISSUE THREAD

**Issue by GitMensch** [spgennard/vscode_cobol#173]
I have a multi-folder workspace (all on NAS), in general most works fine (for the rest more issues are to come).
One think that I've recognized:

* reload window
* COBOL: process workspace for metadata
* run tasks

--> one sees "activating extensions" and then nothing happens (I guess something like a timeout)

What does work:

* reload window
* run tasks (all tasks are shown), press escape
* COBOL: process workspace for metadata
* run tasks

--> all tasks are shown and I can use one (in my case a big one that parses compiler messages from text files for the complete workspace)

Please investigate if one of the following is possible:

* Let "process metadata" (both the global and the "per file" one) be done asynchronously. This potentially also allows vscode to stay responsive in "complex metadata scenarios" (that was reported before, I personally had no issue with that).
* Only if the former is not possible: Trigger whatever is done with the extension when "run tasks" is issued the first time also on the beginning of "process workspace for metadata".

**Comment by spgennard** [comment]
Can you please include your user/workspace settings to understand what features are on or not, as many of the features are already in async blocks.

Global processing of files is either done at startup (if opted in) or via a command.. both of which cannot be done in an async manner.

"per file" parsing is done from the source file in the editor, which is done on-demand on a feature basic, for example: outline view requires parsing, symbol searching requires parsing etc..  all of which you can turn off.

**Comment by GitMensch** [comment]
> Global processing of files is either done at startup (if opted in) or via a command.. both of which cannot be done in an async manner.

It is setup to only be done upon request (and as I've specified this is where the plugin seems to hang the tasks during the process [I'd guess vscode asks all extensions for tasks and this extension is still blocked by the metadata search]), but as I have moved all configuration from ".vscode" to the workspace file now it would be reasonable to set it to be on startup...

Can you please explain why those two can't be done async?

Settings in use:

```json
{
	"folders": [
		{
			"path": ".",
			"name": "root directory"
		},
		{
			"path": "src",
			"name": "src root directory"
		},
		{
			"path": "src/module1",
		},
		{
			"path": "src/module2",
		},
		{
			"path": "src/module3",
		},
		{
			"path": "src/module4",
		},
		{
			"path": "src/module5",
		},
		{
			"path": "src/module6",
		},
		{
			"path": "src/module7",
		},
		{
			"path": "../userhome/me/src",
			"name": "Home-src directory 4 win-client"
		},
		{
			"path": "/home/me/src",
			"name": "Home-src directory"
		}
	],
	"extensions": {
		"recommendations": [
			"bitlang.cobol"
		]
	},
	"settings": {
		"terminal.integrated.fontFamily": "monospace",
		"window.zoomLevel": 0,
		"gulp.autoDetect": "off",
		"files.associations": {
			"*.cpy": "COBOL",
			"*.pco": "COBOL",
			"*.cob": "COBOL"
		},
		"[COBOL]": {
			"files.encoding": "iso885915",
			"files.autoGuessEncoding": false
		},
		"coboleditor.cache_metadata": "on",
		"coboleditor.copybookexts": [
			"cpy",
			"CPY"
		],
		"coboleditor.cache_directory_strategy": "storagepath",
		"coboleditor.copybooks_nested": true,
		"coboleditor.experimential_features": true,
		"coboleditor.fileformat_strategy": "always_fixed",
		"coboleditor.parse_copybooks_for_references": true,
		"coboleditor.enable_auto_tasks": false
	},
	"tasks": {
		"version": "2.0.0",
		"inputs": [
			{
				"id": "exeProgram",
				"type": "promptString",
				"description": "Program to run, leave empty to execute interactively",
				"default": ""
			}
		],
		"tasks": [
			{
				"label": "Build (SSH)",
				"windows": {
					"command": "C:\\Program Files (x86)\\PuTTY\\plink.exe",
					"args": [
						"${env:USERNAME}@cobsrv",
						"-batch",
						"/home/me/build.sh --area=area1 --do=${fileBasenameNoExtension} --path-translate"
					]
				},
				"linux": {
					"command": "ssh",
					"args": [
						"cobsrv",
						"/home/me/build.sh --area=area1 --do=${fileBasenameNoExtension}"
					]
				},
				"presentation": {
					"echo": false,
					"reveal": "always",
					"focus": false,
					"panel": "dedicated",
					"showReuseMessage": false,
					"clear": true
				},
				"problemMatcher": [
					"$gnucobol3-cobc",
					"$mfcobol-errformat2-netx-sx"
				],
				"group": {
					"kind": "build",
					"isDefault": true
				}
			},
			{
				"label": "Show Complete Last State (SSH)",
				"windows": {
					"command": "C:\\Program Files (x86)\\PuTTY\\plink.exe",
					"args": [
						"${env:USERNAME}@cobsrv",
						"-batch",
						"/home/me/build.sh --area=area1 --do=getRes --path-translate"
					]
				},
				"linux": {
					"command": "ssh",
					"args": [
						"cobsrv",
						"/home/me/build.sh --area=area1 --do=getRes"
					]
				},
				"presentation": {
					"echo": false,
					"reveal": "never",
					// "focus": false,
					"panel": "dedicated",
					"showReuseMessage": false,
					"clear": true
				},
				"problemMatcher": [
					"$gnucobol3-cobc",
					"$mfcobol-errformat2-netx-sx"
				],
				"group": "none"
			},
			{
				"label": "Run Program (SSH)",
				"windows": {
					"command": "C:\\Program Files (x86)\\PuTTY\\plink.exe",
					"args": [
						"${env:USERNAME}@cobsrv",
						"-t",
						"doSetup area1 1>/dev/null && echo \"running \"${input:exeProgram} \"on $area\" && echo \"\" && runner ${input:exeProgram}; exit $?"
					]
				},
				"linux": {
					"command": "ssh",
					"args": [
						"cobsrv",
						"doSetup area1 1>/dev/null && echo \"running \"${input:exeProgram} \"on $area\" && echo \"\" && runner ${input:exeProgram}; exit $?"
					]
				},
				"problemMatcher": [],
				"group": "none"
			},
		]
	}
}
```

**Comment by spgennard** [comment]
The use ```"coboleditor.parse_copybooks_for_references": true,``` will cause the parser to do file access 
rather than source access via vscode.

Turning this off will make your life easier.

**Comment by GitMensch** [comment]
> The use `"coboleditor.parse_copybooks_for_references": true,` will cause the parser to do file access
> rather than source access via vscode.

So the copybooks will still be parsed, correct? In this case it may be good to adjust its description or get rid of it if you're sure it isn't needed (maybe it is when not working with a workspace?)

**Comment by GitMensch** [comment]
I've just checked the behavior with `"coboleditor.process_metadata_cache_on_start": true` - currently not usable here as it stands on "Activating Extensions..." for multiple minutes now. Are you sure there is no way to start the metadata parsing asynchronous?

**Comment by spgennard** [comment]
Is this still an issue (as more methods are now being done async)?

**Comment by GitMensch** [comment]
Rechecked, I regularly see that the extensions "hang" (for example getting "extensions are loaded..." with no result when executing "run task" or the outline waiting to gather symbols even for very tiny sources [<28 lines]), but I don't know if bitlang.cobol is to be blamed or not... but I do know that I don't have many extensions manually installed/enabled [this + cobol debugging extension, and I _think_ previously only this one], but there are also the in-built extensions. I try to monitor the process explorer in this case, on thing I've regularly seen there before was:

![vscode process explorer](https://user-images.githubusercontent.com/6699539/92687048-1938e300-f33b-11ea-9735-219deee166ae.png)

So maybe it is the inbuilt json extension...

**Comment by GitMensch** [comment]
OK, got this issue again, but this time:

* opened vscode workspace
* opened file from workspace (via [CTRL]+[P])
* run build task (via [CTRL]+[B])
* got "extensions loading" in the statusbar, clicking on that lead to a popup message that the extension is unresponsive, clicking on details showed:

![grafik](https://user-images.githubusercontent.com/6699539/93359168-93d59700-f842-11ea-9606-ea613da7f764.png)

So I think this is _not_ metadata related, but a general issue. (I guess closing this one and opening a new issue would be reaonsable?)

**Comment by spgennard** [comment]
@GitMensch - Are you still using network drives?

**Comment by GitMensch** [comment]
Yes. The devs operate mainly on win32 while the compilers and run time are on GNU/Linux, the later connecting to a GUI that's on win32 or browser.

Can I somehow log what's happening?

**Comment by spgennard** [comment]
The task is done in a async method, the directories are done in a async promise and sync'ed for completion (to save the cache).

I don't believe I can do anything extra to resolve your network performance related issues.

**Comment by spgennard** [comment]
The settings "coboleditor.cache_metadata_time_limit" has been introduced to allow the user to control how much time can be taken to process the meta data on startup.

The user now has a mechanism to prevent the "halt", so the issue can be considered resolved.

Long term, the dependency on the metadata cache is being reduced and because of this, the impact over time will also be reduced.
